zcash_client_backend/src/zip321.rs

```diff
@@ -0,0 +1,1046 @@
+            .map_err(MemoError::InvalidBase64)
+            .and_then(|b| RawMemo::from_bytes(&b))
+    }
+}
+
+impl Debug for RawMemo {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        f.debug_struct("RawMemo")
+            .field("memo", &format!("{:?}...", &self.0[0..17]))
+            .finish()
+    }
+}
+
+impl PartialEq for RawMemo {
+    fn eq(&self, other: &Self) -> bool {
+        self.to_base64() == other.to_base64()
+    }
+}
+
+impl FromStr for RawMemo {
+    type Err = MemoError;
+
+    fn from_str(memo: &str) -> Result<Self, Self::Err> {
+        RawMemo::from_str(memo)
+    }
+}
+
+impl Eq for RawMemo {}
+
+impl PartialOrd for RawMemo {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.to_base64().cmp(&other.to_base64()))
+    }
+}
+
+impl Ord for RawMemo {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+// RawMemo is somewhat duplicative of the `Memo` type
+// in crate::note_encryption but as that's actively being
+// updated at time of this writing, these functions provide
+// shims to ease future use of those
+pub fn memo_from_vec(v: &[u8]) -> Result<RawMemo, MemoError> {
+    RawMemo::from_bytes(v)
+}
+
+pub fn memo_to_base64(memo: &RawMemo) -> String {
+    memo.to_base64()
+}
+
+pub fn memo_from_base64(s: &str) -> Result<RawMemo, MemoError> {
+    RawMemo::from_base64(s)
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Zip321Payment {
+    recipient_address: Address,
+    amount: Amount,
+    memo: Option<RawMemo>,
+    label: Option<String>,
+    message: Option<String>,
+    other_params: Vec<(String, String)>,
+}
+
+impl Zip321Payment {
+    #[cfg(any(test, feature = "test-dependencies"))]
+    pub(in crate::zip321) fn normalize(&mut self) {
+        self.other_params.sort();
+    }
+
+    /// Returns a function which compares two normalized payments, with addresses sorted by their
+    /// string representation given the specified network. This does not perform normalization
+    /// internally, so payments must be normalized prior to being passed to the comparison function
+    /// returned from this method.
+    #[cfg(any(test, feature = "test-dependencies"))]
+    pub(in crate::zip321) fn compare_normalized<'a, P: consensus::Parameters>(
+        params: &'a P,
+    ) -> impl Fn(&Zip321Payment, &Zip321Payment) -> Ordering + 'a {
+        move |a: &Zip321Payment, b: &Zip321Payment| {
+            let a_addr = a.recipient_address.render(params);
+            let b_addr = b.recipient_address.render(params);
+
+            a_addr
+                .cmp(&b_addr)
+                .then(a.amount.cmp(&b.amount))
+                .then(a.memo.cmp(&b.memo))
+                .then(a.label.cmp(&b.label))
+                .then(a.message.cmp(&b.message))
+                .then(a.other_params.cmp(&b.other_params))
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Zip321Request {
```

Comment:
```suggestion
pub struct PaymentRequest {
```

Reply:
Add documentation to (what are currently named) `Zip321Request` and `Zip321Payment`.

The other `pub struct`s and `pub enum`s will be refactored out of existence, so documenting them right now is non-blocking.